### PR TITLE
Fix jclouds sshable

### PR DESCRIPTION
**Pull request by aledsage**
(Unrelated to other changes, fixes `NetworkingUtilsTest.testIsPortAvailableReportsTrueWhenPortIsFree`).

Lots of fixes to get the `JcloudsLoginLiveTest` working again.

- registerJcloudsSshMachineLocation: use the userCredentials that are passed into the method, rather than looking up what we expect from the config.  
  (Hopefully fixes testSpecifyingPasswordAndSshKeysPrefersKeys).
- Fix waitForSshable to when password supplied on JcloudsLocation config, and using useJcloudsSshInit=false.
- Fix use of DISABLE_ROOT_AND_PASSWORD_SSH: if for root user with password-based login, then explicitly set PasswordAuthentication=yes.  
  If root user, then explicitly set PermitRootLogin=true
- Fix getPublicHostnameAws so pass in the credentials (some code-paths will not have configured the credentials in the generic config).  
  Same for getPrivateHostnameAws.
- Tidy up the createTemporarySshMachineLocaton
- Logging of timing: include how long to get semaphore
- Logging of failure: include how long for each stage
    
Previously hit a problem when the jclouds location was configured with
a password + a blank ssh key, and using useJcloudsSshInit=false.
It failed to ssh to the newly provisioned AWS VM in waitForSshable. It
got the node.getCredentials, which gives us the ssh-key from the AWS
key-pair. But then we instantiate an SshMachineLocation to check if we
can connection (before continuing with the user-setup). That
machineLocation inherits the password (which has not been set on the
machine), so it fails to ssh - the password takes precedence over the
ssh-key.

### CODE REVIEW

File: utils/common/src/main/java/org/apache/brooklyn/util/net/Networking.java [hunk 1]
```diff
@@ -486,6 +487,12 @@ public static boolean isLocalhost(String remoteAddress) {
     }
     
     public static boolean isReachable(HostAndPort endpoint) {
+        // TODO Should we create an unconnected socket, and then use the calls below (see jclouds' InetSocketAddressConnect):
+        //      socket.setReuseAddress(false);
+        //      socket.setSoLinger(false, 1);
+        //      socket.setSoTimeout(timeout);
+        //      socket.connect(socketAddress, timeout);
```
Comment (neykov) [inline utils/common/src/main/java/org/apache/brooklyn/util/net/Networking.java]: Don't see the point of the above, but +1 for the `,conntect` timeout.

File: utils/common/src/test/java/org/apache/brooklyn/util/net/ReachableSocketFinderTest.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+            fail("Expected failure, but got "+result);
+        } catch (NoSuchElementException e) {
+            // success
+        }
+    }
+
+    private ServerSocket connectToPort() throws Exception {
+        int port = findAvailablePort();
+        LOG.info("acquired port on "+port+" for test "+JavaClassNames.niceClassAndMethod());
+        return new ServerSocket(port);
```
Comment (neykov) [inline utils/common/src/test/java/org/apache/brooklyn/util/net/ReachableSocketFinderTest.java]: Could use port 0 to avoid someone else stealing the port.

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java [hunk 3]
```diff
@@ -1653,8 +1662,10 @@ protected LoginCredentials createUser(ComputeService computeService, NodeMetadat
                         LOG.warn("exit code {} when creating user for {}; usage may subsequently fail", exitcode, node);
                     }
                 } finally {
-                    getManagementContext().getLocationManager().unmanage(sshLoc);
-                    Streams.closeQuietly(sshLoc);
+                    if (sshLoc != null) {
```
Comment (neykov) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java]: Can't be null

**Comment by neykov** [comment]
Good to merge after addressing the potential NPE error, others are minor.

File: utils/common/src/main/java/org/apache/brooklyn/util/net/Networking.java (hunk 1 above)
Comment (aledsage) [inline utils/common/src/main/java/org/apache/brooklyn/util/net/Networking.java]: Let's do the actual change in a different PR.

**Comment by aledsage** [comment]
Thanks @neykov - merging now.

I didn't address your comment "Will be able to get rid of compoundFuture as well, leaving out the need for ListeningExecutorService (and could even use Brooklyn's task manager)." We could look at that in a future PR, but I'm not quite sure what you're proposing.
